refactoring of coverage for future additions

jamestwebber opened the pull request:
I mostly refactored this just so I understood how it worked, but I think this is slightly cleaner. And it still supports massive regions, although nobody should use it for that.

Commit:
fixed overlapping track bug, removed redundant points for speed

Commit:
haha whoops

Commit:
add tag-based and peak-based coverage tracks

Commit:
added tag options to plot_exons

Commit:
added coverage_tag options to view options

Commit:
fixed bug in single coverage

Commit:
fixed coverage logic for huge/sparse regions

Commit:
refactor single-coverage to match

Commit:
sparse coverage to handle huge regions

Commit:
revamping stacked coverage for future changes

GeorgescuC committed:
Merge branch 'master' into stacked-coverage

Commit:
Make coloring of coverage layers be consistant by starting from the "front".
Optimize coverage series calculation to remove redundant points and reduce file size.

GeorgescuC commented on genomeview/bamtrack.py:
```diff
@@ -744,179 +747,172 @@ def layout(self, scale):
+
+        for read in self._get_reads(scale):
+            if not read.has_tag(self.tag):
+                continue
```
Should reads without the tag be grouped in an extra bin, optionally?

jamestwebber replied:
In my use cases I don't think that'd be useful, but maybe in other situations.

GeorgescuC replied:
Made a change so that they are grouped in a separate secondary bin that gets added over all other layers and uses a different color palette.

Commit:
Link peak coverage binning option

Commit:
Add a secondary_coverage to method that adds the coverage that can be used to stack on top of the regular coverge with a different set of colors, currently used for reads that start before the displayed window for size based bins, for reads that don't fall within any peak bin, or for reads missing a tag.
Reads that are at a distance equal to min_dist in peak binning are not added to that bin anymore to be consistant with peaks definition and since 0 disables the option, 1 is the smallest value allowed and thus should be for strict bins.

Commit:
Fix last step of the coverage track missing

Commit:
clean up comments

Commit:
Make MAX_BINS for peak coverage be counted after merging bins that are too close together.
Fix peak coverage to respect priming orientation/strand for the ordered they are layered in.
